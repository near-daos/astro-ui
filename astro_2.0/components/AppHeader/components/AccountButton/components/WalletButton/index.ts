export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletButton/WalletButton';
