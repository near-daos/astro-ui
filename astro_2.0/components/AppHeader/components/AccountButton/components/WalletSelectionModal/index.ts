export * from 'astro_2.0/components/AppHeader/components/AccountButton/components/WalletSelectionModal/WalletSelectionModal';
