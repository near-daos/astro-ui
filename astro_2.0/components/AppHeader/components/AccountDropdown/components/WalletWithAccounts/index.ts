export * from 'astro_2.0/components/AppHeader/components/AccountDropdown/components/WalletWithAccounts/WalletWithAccounts';
