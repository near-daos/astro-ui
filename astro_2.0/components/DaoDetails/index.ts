export * from './DaoDetailsMinimized';
export * from './DaoDetailsGrid';
