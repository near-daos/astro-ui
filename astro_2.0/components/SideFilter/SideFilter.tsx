import cn from 'classnames';
import Link from 'next/link';
import { useRouter } from 'next/router';
import findIndex from 'lodash/findIndex';
import React, { CSSProperties, useEffect } from 'react';
import { useTranslation } from 'next-i18next';

import styles from './SideFilter.module.scss';

export type ListItem = {
  value?: string;
  label: string;
  disabled?: boolean;
};

interface SideFilterProps {
  title: string;
  disabled?: boolean;
  forceHorizontalView?: boolean;
  queryName: string;
  className?: string;
  titleClassName?: string;
  list?: ListItem[];
  hideAllOption?: boolean;
  shallowUpdate?: boolean;
  itemClassName?: string;
  markerOffset?: number;
}

export const SideFilter = ({
  list,
  title,
  disabled,
  queryName,
  className,
  titleClassName,
  itemClassName,
  hideAllOption = false,
  shallowUpdate = false,
  forceHorizontalView = false,
  markerOffset = 0,
}: SideFilterProps): JSX.Element => {
  const { t } = useTranslation();

  const ITEM_ALL = {
    value: undefined,
    label: t('feed.filters.all'),
  };

  const { query } = useRouter();
  const { [queryName]: value } = query;

  function getActiveMarkTransformValue() {
    const HUNDRED_PERCENT = 100;

    const selectedValue = value as string;
    const index = findIndex(list, { value: selectedValue });

    const transformVal =
      index === -1
        ? markerOffset
        : index * HUNDRED_PERCENT + (hideAllOption ? 0 : HUNDRED_PERCENT);

    return {
      '--filter-active-transform': `${transformVal}%`,
    } as CSSProperties;
  }

  function renderFilterItem(item: ListItem) {
    const { value: itemVal, label, disabled: disabledItem } = item;

    const href = {
      query: {
        ...query,
        [queryName]: itemVal,
      },
    };

    const hrefClassName = cn(
      styles.categoriesListItem,
      {
        [styles.categoriesListItemSelected]:
          value === itemVal || (!itemVal && !value),
        [styles.disabled]: disabled || disabledItem,
      },
      itemClassName
    );

    return (
      <li className={styles.categoriesListItemWrapper} key={label}>
        <Link href={href} replace scroll={false} shallow={shallowUpdate}>
          <a className={hrefClassName} tabIndex={disabled ? -1 : 0}>
            {label}
          </a>
        </Link>
      </li>
    );
  }

  useEffect(() => {
    const selectedItem = document.querySelector(
      `.${styles.categoriesListItemSelected}`
    );

    if (selectedItem) {
      selectedItem.scrollIntoView({
        behavior: 'smooth',
        block: 'nearest',
        inline: 'center',
      });
    }
  }, [value]);

  return (
    <div
      className={cn(styles.categoriesListRoot, className, {
        [styles.categoriesListHorizontal]: forceHorizontalView,
      })}
    >
      <p className={cn(styles.categoriesListTitle, titleClassName)}>{title}</p>

      <ul
        className={styles.categoriesList}
        style={getActiveMarkTransformValue()}
      >
        {!hideAllOption && renderFilterItem(ITEM_ALL)}
        {list?.map(renderFilterItem)}
      </ul>
    </div>
  );
};
