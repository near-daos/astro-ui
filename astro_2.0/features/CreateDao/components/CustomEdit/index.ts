import CustomEdit from './CustomEdit';

export default CustomEdit;
