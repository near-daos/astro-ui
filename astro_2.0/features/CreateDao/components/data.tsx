import {
  DAOProposalsType,
  DaoSettingOption,
  DAOStructureType,
  DAOVotingPowerType,
  DaoSubjectInfo,
} from 'astro_2.0/features/CreateDao/components/types';

const DAO_PROPOSALS_OPEN: DaoSettingOption<DAOProposalsType> = {
  value: 'open',
  icon: 'illustrationOpenOrganization',
  subject: 'proposals',
  title: 'daoRulesAll',
  description: 'daoRulesAllDescription',
};

const DAO_PROPOSALS_CLOSED: DaoSettingOption<DAOProposalsType> = {
  value: 'closed',
  icon: 'illustrationMembersOnly',
  subject: 'proposals',
  title: 'daoRulesMembers',
  description: 'daoRulesMembersDescription',
};

const DAO_VOTING_POWER_DEMOCRATIC: DaoSettingOption<DAOVotingPowerType> = {
  value: 'democratic',
  icon: 'illustrationVotePerMember',
  subject: 'voting',
  title: 'daoRulesDemocratic',
  description: 'daoRulesDemocraticDescription',
};

const DAO_VOTING_POWER_TOKEN: DaoSettingOption<DAOVotingPowerType> = {
  value: 'weighted',
  icon: 'illustrationTokenWeighted',
  subject: 'voting',
  title: 'daoRulesTokenWeighted',
  description: 'daoRulesTokenWeightedDescription',
};

const DAO_STRUCTURE_FLAT: DaoSettingOption<DAOStructureType> = {
  value: 'flat',
  icon: 'illustrationFlatOrganization',
  subject: 'structure',
  title: 'daoRulesFlat',
  description: 'daoRulesFlatDescription',
};

const DAO_STRUCTURE_GROUPS: DaoSettingOption<DAOStructureType> = {
  value: 'groups',
  icon: 'illustrationGroupsAndCommitttees',
  subject: 'structure',
  title: 'daoRulesGroups',
  description: 'daoRulesGroupsDescription',
};

export const DAO_PROPOSALS_OPTIONS: Record<
  DAOProposalsType,
  DaoSettingOption<DAOProposalsType>
> = {
  open: DAO_PROPOSALS_OPEN,
  closed: DAO_PROPOSALS_CLOSED,
};

export const DAO_VOTING_POWER_OPTIONS: Record<
  DAOVotingPowerType,
  DaoSettingOption<DAOVotingPowerType>
> = {
  democratic: DAO_VOTING_POWER_DEMOCRATIC,
  weighted: DAO_VOTING_POWER_TOKEN,
};

export const DAO_STRUCTURE_OPTIONS: Record<
  DAOStructureType,
  DaoSettingOption<DAOStructureType>
> = {
  flat: DAO_STRUCTURE_FLAT,
  groups: DAO_STRUCTURE_GROUPS,
};

export const DAO_SUBJECT_OPTIONS = {
  voting: Object.values(DAO_VOTING_POWER_OPTIONS),
  structure: Object.values(DAO_STRUCTURE_OPTIONS),
  proposals: Object.values(DAO_PROPOSALS_OPTIONS),
};

export const DAO_RULES_INFO: DaoSubjectInfo[] = [
  {
    subject: 'proposals',
    title: 'daoRulesProposals',
    subTitle: 'daoRulesProposalsDescription',
  },
  {
    subject: 'voting',
    title: 'daoRulesVoting',
    subTitle: 'daoRulesVotingDescription',
  },
  {
    subject: 'structure',
    title: 'daoRulesStructure',
    subTitle: 'daoRulesStructureDescription',
  },
];
