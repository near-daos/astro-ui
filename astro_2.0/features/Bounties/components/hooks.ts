import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
