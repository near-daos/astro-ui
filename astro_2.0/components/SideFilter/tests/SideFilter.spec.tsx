/* eslint-disable @typescript-eslint/ban-ts-comment */

import { render } from 'jest/testUtils';
import { useRouter } from 'next/router';

import { SideFilter } from 'astro_2.0/components/SideFilter';

jest.mock('next/router', () => {
  return {
    useRouter: jest.fn().mockImplementation(() => ({
      query: {},
    })),
  };
});

beforeAll(() => {
  window.HTMLElement.prototype.scrollIntoView = jest.fn();
});

describe('Side filter', () => {
  it('Should render component', () => {
    const { container } = render(
      <SideFilter title="Hello World!" queryName="" />
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render list item', () => {
    // @ts-ignore
    useRouter.mockImplementation(() => ({
      query: {
        name: '1',
      },
    }));

    const listItemLabel = 'Hey!';
    const { getAllByText } = render(
      <SideFilter
        title="Hello World!"
        queryName="name"
        list={[{ label: listItemLabel, value: '1' }]}
      />
    );

    expect(getAllByText(listItemLabel)).toHaveLength(1);
  });
});

/* eslint-enable @typescript-eslint/ban-ts-comment */
