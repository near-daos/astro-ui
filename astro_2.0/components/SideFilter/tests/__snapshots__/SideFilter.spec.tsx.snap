// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Side filter Should render component 1`] = `
<div>
  <div
    class="categoriesListRoot"
  >
    <p
      class="categoriesListTitle"
    >
      Hello World!
    </p>
    <ul
      class="categoriesList"
      style="--filter-active-transform: 0%;"
    >
      <li
        class="categoriesListItemWrapper"
      >
        <a
          class="categoriesListItem categoriesListItemSelected"
          href="/?="
          tabindex="0"
        >
          feed.filters.all
        </a>
      </li>
    </ul>
  </div>
</div>
`;
